PRINCIPLES.md

```diff
@@ -0,0 +1,254 @@
+
+This relative readability also implies that the users are relevant when evaluating whether a system follows this principle. For example, a Desired State defined with a rich grammar of S-Expressions would suite a team of developers with a background in functional programming, but would violate this principle if the majority of the team found the format incomprehensible.
+
+Having a human-readable Desired State does not in any way preclude the use of rich tooling or graphical interfaces that facilitate Desired State generation and interpretation. It only require that the canonical source of truth be human readable and writable plain text.
+
+#### How much of a system must be declared?
+
+Ideally, all of it; and the entire system can be recreated exclusively from its Desired State.
+
+That said, the definition of a "computer systems" can be quite broad and strays into human processes and systems as well. 
+For example, is a company's sales process a "computer system"? 
+Likely in parts. 
+Should GitOps be applied to the entire sales process including the human processes (without prejudice as to what those processes are)? 
+Whilst such a vision is compelling, a more pragmatic approach is preferrable, otherwise we risk trying to boil the ocean.
+
+Instead, we should focus on subsystems where the Desired State is well defined, implement the GitOps principles there, and grow out to capture more systems from that initial subsystem. <!-- This progressive approach to GitOps is elaborated further in [the guide to adopting GitOps.md](ADOPTING_GITOPS.md). -->
+
+### 2. Immutable configuration versions
+
+<div style="background: #E3F2FD; border-left: 3px solid #0D47A1; padding:10px; margin-bottom:5px; font-style:italic">
+<em>Desired State</em> is stored in a way that supports versioning, immutability of versions, and retains a complete version history.
+We call systems that store Desired State in this way <em>State Stores</em>.
+</div>
+
+#### What forms a version?
+
+A version is the Desired State for a system as a whole. It is the canonical form of what we desire the system to be at a point in time.
+
+It is insufficient to version part of the Desired State or to version these parts in separate State Stores. Real software systems often have overarching behaviour that is the result of coupling between components. If the Desired State of these components were to change independently, it would be difficult to map a change in observed behaviour of our system to a single change in Desired State. Being able to make this 1:1 mapping is operationally benefitial, as we can then map behavioural issues of our system directly to the changes that occured. The utility of having the entire system defined in a single canonical location grows in proportion to the complexity and internal coupling of the system. A web of references to configuration data located in different locations is undesirable, as it makes understanding the desired state particularly difficult.
```

Comment:
Not sure I follow this in the context of application configuration vs platform configuration as desired state. For instance, I may have some state in which I hold the declaration that I need some service running or crd available in which my application will depend (and maybe other applications as well), and some other state in which I specify the runtime parameters for one such specific application (say, endpoints or other runtime specific configuration for the application itself) - not sure its the best example but the point is, as part of a system there may be platform components as well as application components that will be interrelated, and following this statements it seems the argument is, anything that is somehow interrelated should be declared in the same state store. I think that such interpretation would be in the least not realistic.

Also I'd change `real software systems` to `production software systems` or `distributed systems`, can't think of a surreal or imaginary software system 😆

Reply:
😄 Indeed.

Reply:
> the point is, as part of a system there may be platform components as well as application components that will be interrelated, and following this statements it seems the argument is, anything that is somehow interrelated should be declared in the same state store. I think that such interpretation would be in the least not realistic.

I agree that for all practical purposes, this is currently unrealistic, but the general point still stands. It is desirable to have all of a system's configuration in a single state store. I think changing "insufficient" to "undesirable" may be better. Do you think that's sufficient @murillodigital?

Reply:
> 
> 
> > the point is, as part of a system there may be platform components as well as application components that will be interrelated, and following this statements it seems the argument is, anything that is somehow interrelated should be declared in the same state store. I think that such interpretation would be in the least not realistic.
> 
> I agree that for all practical purposes, this is currently unrealistic, but the general point still stands. It is desirable to have all of a system's configuration in a single state store. I think changing "insufficient" to "undesirable" may be better. Do you think that's sufficient @murillodigital?

How about changing the approach from that which is undesirable or suboptimal to that which is ideal:

`Ideally, the configuration for all interrelated components of a system should be stored in a single state store, in order to have a concise and consolidate view across all components that may have be subject to overarching behavior.`

Or something along those lines? @bricef

Reply:
> How about changing the approach from that which is undesirable or suboptimal to that which is ideal

@murillodigital I like this approach flip